symengine/sets.cpp

```diff
@@ -464,4 +559,253 @@ RCP<const Boolean> Union::contains(const RCP<const Basic> &o) const
+            othersets.insert(input);
+    }
+    if (fsets.size() != 0) {
+        const FiniteSet &fs = down_cast<const FiniteSet &>(**fsets.begin());
+        auto cont = fs.get_container();
+        fsets.erase(fsets.begin());
+        set_basic finalfs;
+        for (const auto &fselement : cont) {
+            bool present = true;
+            for (const auto &fset : fsets) {
+                present
+                    = present and eq(*(fset->contains(fselement)), *boolTrue);
+            }
+            if (!present)
+                continue;
+            for (const auto &oset : othersets) {
+                present
+                    = present and eq(*(oset->contains(fselement)), *boolTrue);
+            }
+            if (present)
+                finalfs.insert(fselement);
+        }
+        return finiteset(finalfs);
+    }
+
+    // If any of the sets is union, then return a Union of Intersections
+    for (auto it = incopy.begin(); it != incopy.end(); ++it) {
+        if (is_a<Union>(**it)) {
+            auto container = down_cast<const Union &>(**it).get_container();
+            incopy.erase(it);
+            auto other = SymEngine::set_intersection(incopy);
+            set_set usets;
+            for (const auto &c : container) {
+                usets.insert(SymEngine::set_intersection({c, other}));
+            }
+            return SymEngine::set_union(usets);
+        }
+    }
+
+    // Simplify and return a `Complement` if any of the sets is a complement
+    for (auto it = incopy.begin(); it != incopy.end(); ++it) {
+        if (is_a<Complement>(**it)) {
+            auto container
+                = down_cast<const Complement &>(**it).get_container();
+            auto universe = down_cast<const Complement &>(**it).get_universe();
+            incopy.erase(it);
+            incopy.insert(universe);
+            auto other = SymEngine::set_intersection(incopy);
+            return SymEngine::set_complement(other, container);
+        }
+    }
+    // Pair-wise rules
+    bool shouldContinue = true;
+    while (shouldContinue) {
+        for (const auto &input : incopy) {
+            shouldContinue = false;
+            set_set other_sets = incopy;
+            auto it = other_sets.find(input);
+            other_sets.erase(it);
+            set_set newinput;
+            for (auto &oset : other_sets) {
+                // TO-DO: needs the following improvement once Intersection
+                // class is implemented.
+                // input_oset if found to be not simplified, then skip this
+                // pair.
+                try {
+                    auto input_oset = input->set_intersection(oset);
+                    newinput = other_sets;
+                    it = newinput.find(oset);
+                    newinput.erase(it);
+                    newinput.insert(input_oset);
+                    shouldContinue = true;
+                    break;
+                } catch (std::runtime_error &err) {
```

Comment:
`input->set_intersection(oset)` returns "Not implemented" runtime_error, if it can't do intersection between `input` and `oset`.  
Later when `Intersection` class gets implemented, this can be changed easily to check if there is any simplification as mentioned above in the TO-DO comment.